File: packages/core-ts/src/policy.ts
import { readFile } from 'node:fs/promises';
import type {
  Policy,
  PolicyResult,
  PolicyDecision,
  AgentManifest,
  TrustLevel,
} from './types.js';

export interface PolicyContext {
  manifest: AgentManifest;
  trustLevel: TrustLevel;
  action: string;
  actionParams?: Record<string, unknown>;
  revoked: boolean;
  expired: boolean;
}

/** Evaluate a request against a policy. Returns a decision with reasons. */
export function evaluatePolicy(context: PolicyContext, policy: Policy): PolicyResult {
  const matchingRules = policy.rules.filter((rule) => rule.action === context.action);

  if (matchingRules.length === 0) {
    return {
      decision: policy.default,
      reasons: [`No policy rule matches action "${context.action}"; applying default: ${policy.default}`],
    };
  }

  for (const rule of matchingRules) {
    const failures: string[] = [];

    if (rule.require.manifest_not_expired && context.expired) {
      failures.push('Manifest is expired');
    }

    if (rule.require.not_revoked && context.revoked) {
      failures.push('Agent key has been revoked');
    }

    if (
      rule.require.min_trust_level !== undefined &&
      context.trustLevel < rule.require.min_trust_level
    ) {
      failures.push(
        `Trust level ${context.trustLevel} is below required ${rule.require.min_trust_level}`
      );
    }

    if (rule.require.capability_declared) {
      const hasCap = context.manifest.capabilities.some(
        (c) => c.id === rule.require.capability_declared
      );
      if (!hasCap) {
        failures.push(
          `Required capability "${rule.require.capability_declared}" not declared in manifest`
        );
      }
    }

    if (rule.require.constraints_satisfied && context.actionParams) {
      const constraintFailures = checkConstraints(context);
      failures.push(...constraintFailures);
    }

    if (failures.length > 0) {
      const reasons = failures.map((f) => {
        if (rule.reason_template) {
          return rule.reason_template
            .replace('{{agent_id}}', context.manifest.agent_id)
            .replace('{{failed_check}}', f);
        }
        return f;
      });

      return {
        decision: rule.on_fail,
        reasons,
        matched_rule: rule.action,
      };
    }
  }

  return {
    decision: 'ACCEPT' as PolicyDecision,
    reasons: ['All policy checks passed'],
    matched_rule: matchingRules[0].action,
  };
}

/** Load a policy from a local file path or URL. */
export async function loadPolicy(source: string): Promise<Policy> {
  let raw: string;

  if (source.startsWith('http://') || source.startsWith('https://')) {
    const response = await fetch(source);
    if (!response.ok) {
      throw new Error(`Failed to fetch policy from ${source}: ${response.status}`);
    }
    raw = await response.text();
  } else {
    raw = await readFile(source, 'utf-8');
  }

  return JSON.parse(raw) as Policy;
}

function checkConstraints(context: PolicyContext): string[] {
  const failures: string[] = [];
  const params = context.actionParams ?? {};

  const capability = context.manifest.capabilities.find((c) => c.id === context.action);
  if (!capability?.constraints) return failures;

  const constraints = capability.constraints;

  if (
    constraints.max_amount_usd !== undefined &&
    typeof params.amount_usd === 'number' &&
    params.amount_usd > constraints.max_amount_usd
  ) {
    failures.push(
      `Amount $${params.amount_usd} exceeds maximum allowed $${constraints.max_amount_usd}`
    );
  }

  if (
    constraints.require_human_approval_above_usd !== undefined &&
    typeof params.amount_usd === 'number' &&
    params.amount_usd > constraints.require_human_approval_above_usd
  ) {
    failures.push(
      `Amount $${params.amount_usd} exceeds human approval threshold $${constraints.require_human_approval_above_usd}`
    );
  }

  return failures;
}

File: packages/core-py/agentid/policy.py
"""Policy evaluation engine."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Optional

from agentid.types import (
    AgentManifest,
    Policy,
    PolicyDecision,
    PolicyResult,
    PolicyRule,
    TrustLevel,
)


def evaluate_policy(
    *,
    manifest: AgentManifest,
    trust_level: TrustLevel,
    action: str,
    action_params: Optional[dict[str, Any]] = None,
    revoked: bool = False,
    expired: bool = False,
    policy: Policy,
) -> PolicyResult:
    """Evaluate a request against a policy. Returns a decision with reasons."""
    matching_rules = [r for r in policy.rules if r.action == action]

    if not matching_rules:
        return PolicyResult(
            decision=PolicyDecision(policy.default),
            reasons=[f'No policy rule matches action "{action}"; applying default: {policy.default}'],
        )

    for rule in matching_rules:
        failures: list[str] = []
        req = rule.require

        if req.get("manifest_not_expired") and expired:
            failures.append("Manifest is expired")

        if req.get("not_revoked") and revoked:
            failures.append("Agent key has been revoked")

        min_trust = req.get("min_trust_level")
        if min_trust is not None and trust_level.value < min_trust:
            failures.append(
                f"Trust level {trust_level.value} is below required {min_trust}"
            )

        cap_required = req.get("capability_declared")
        if cap_required:
            has_cap = any(c.id == cap_required for c in manifest.capabilities)
            if not has_cap:
                failures.append(
                    f'Required capability "{cap_required}" not declared in manifest'
                )

        if req.get("constraints_satisfied") and action_params:
            failures.extend(_check_constraints(manifest, action, action_params))

        if failures:
            reasons = []
            for f in failures:
                if rule.reason_template:
                    reasons.append(
                        rule.reason_template
                        .replace("{{agent_id}}", manifest.agent_id)
                        .replace("{{failed_check}}", f)
                    )
                else:
                    reasons.append(f)

            return PolicyResult(
                decision=PolicyDecision(rule.on_fail),
                reasons=reasons,
                matched_rule=rule.action,
            )

    return PolicyResult(
        decision=PolicyDecision.ACCEPT,
        reasons=["All policy checks passed"],
        matched_rule=matching_rules[0].action,
    )


def load_policy(source: str) -> Policy:
    """Load a policy from a file path."""
    raw = Path(source).read_text(encoding="utf-8")
    data = json.loads(raw)

    rules = [
        PolicyRule(
            action=r["action"],
            require=r["require"],
            on_fail=r["on_fail"],
            reason_template=r.get("reason_template"),
        )
        for r in data["rules"]
    ]

    return Policy(
        policy_version=data["policy_version"],
        rules=rules,
        default=data["default"],
    )


def _check_constraints(
    manifest: AgentManifest, action: str, params: dict[str, Any]
) -> list[str]:
    failures: list[str] = []

    capability = next((c for c in manifest.capabilities if c.id == action), None)
    if not capability or not capability.constraints:
        return failures

    constraints = capability.constraints

    if (
        constraints.max_amount_usd is not None
        and isinstance(params.get("amount_usd"), (int, float))
        and params["amount_usd"] > constraints.max_amount_usd
    ):
        failures.append(
            f"Amount ${params['amount_usd']} exceeds maximum allowed ${constraints.max_amount_usd}"
        )

    if (
        constraints.require_human_approval_above_usd is not None
        and isinstance(params.get("amount_usd"), (int, float))
        and params["amount_usd"] > constraints.require_human_approval_above_usd
    ):
        failures.append(
            f"Amount ${params['amount_usd']} exceeds human approval threshold "
            f"${constraints.require_human_approval_above_usd}"
        )

    return failures
